File: Packages/com.gamelovers.mcp-unity/Server~/src/tools/setPlayModeStatusTool.ts
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { McpUnity } from '../unity/mcpUnity.js';
import { McpUnityError, ErrorType } from '../utils/errors.js';
import * as z from 'zod';
import { Logger } from '../utils/logger.js';
import { CallToolResult } from '@modelcontextprotocol/sdk/types.js';

const toolName = 'set_play_mode_status';
const toolDescription = "Controls Unity play mode. Actions: 'play' (start or unpause), 'pause' (toggle pause), 'stop' (exit play mode), 'step' (advance one frame while paused).";

const paramsSchema = z.object({
  action: z.enum(['play', 'pause', 'stop', 'step']).describe("The play mode action to execute: 'play', 'pause', 'stop', or 'step'")
});

export function registerSetPlayModeStatusTool(server: McpServer, mcpUnity: McpUnity, logger: Logger) {
  logger.info(`Registering tool: ${toolName}`);

  server.tool(
    toolName,
    toolDescription,
    paramsSchema.shape,
    async (params: any) => {
      try {
        logger.info(`Executing tool: ${toolName}`, params);
        const result = await toolHandler(mcpUnity, params);
        logger.info(`Tool execution successful: ${toolName}`);
        return result;
      } catch (error) {
        logger.error(`Tool execution failed: ${toolName}`, error);
        throw error;
      }
    }
  );
}

async function toolHandler(mcpUnity: McpUnity, params: any): Promise<CallToolResult> {
  const validatedParams = paramsSchema.parse(params);
  
  const response = await mcpUnity.sendRequest({
    method: toolName,
    params: validatedParams
  });

  if (!response.success) {
    throw new McpUnityError(
      ErrorType.TOOL_EXECUTION,
      response.message || `Failed to execute play mode action: ${validatedParams.action}`
    );
  }

  const statusText = response.isPlaying
    ? (response.isPaused ? 'Playing (paused)' : 'Playing')
    : 'Edit mode';

  return {
    content: [
      {
        type: 'text',
        text: `Play mode action '${validatedParams.action}' executed successfully. Current state: ${statusText}`
      }
    ],
    data: {
      action: validatedParams.action,
      isPlaying: response.isPlaying,
      isPaused: response.isPaused
    },
    isError: false
  };
}

File: Packages/com.gamelovers.mcp-unity/Editor/Tools/SetPlayModeStatusTool.cs
using System;
using Newtonsoft.Json.Linq;
using UnityEditor;
using McpUnity.Unity;
using McpUnity.Utils;

namespace McpUnity.Tools
{
    /// <summary>
    /// Tool for controlling Unity play mode (play, pause, step)
    /// </summary>
    public class SetPlayModeStatusTool : McpToolBase
    {
        public SetPlayModeStatusTool()
        {
            Name = "set_play_mode_status";
            Description = "Controls Unity play mode. Actions: 'play', 'pause', 'stop', 'step'.";
        }

        public override JObject Execute(JObject parameters)
        {
            try
            {
                string action = parameters?["action"]?.ToString()?.ToLowerInvariant();
                
                if (string.IsNullOrEmpty(action))
                {
                    return McpUnitySocketHandler.CreateErrorResponse(
                        "Missing required parameter 'action'. Valid actions: 'play', 'pause', 'stop', 'step'",
                        "missing_parameter"
                    );
                }

                bool wasPlaying = EditorApplication.isPlaying;
                bool wasPaused = EditorApplication.isPaused;

                switch (action)
                {
                    case "play":
                        if (!EditorApplication.isPlaying)
                        {
                            // Start play mode
                            EditorApplication.isPlaying = true;
                        }
                        else if (EditorApplication.isPaused)
                        {
                            // Unpause if already playing
                            EditorApplication.isPaused = false;
                        }
                        break;

                    case "pause":
                        if (EditorApplication.isPlaying)
                        {
                            EditorApplication.isPaused = !EditorApplication.isPaused;
                        }
                        else
                        {
                            return McpUnitySocketHandler.CreateErrorResponse(
                                "Cannot pause: Editor is not in play mode",
                                "invalid_state"
                            );
                        }
                        break;

                    case "stop":
                        if (EditorApplication.isPlaying)
                        {
                            EditorApplication.isPlaying = false;
                        }
                        break;

                    case "step":
                        if (EditorApplication.isPlaying)
                        {
                            EditorApplication.Step();
                        }
                        else
                        {
                            return McpUnitySocketHandler.CreateErrorResponse(
                                "Cannot step: Editor is not in play mode",
                                "invalid_state"
                            );
                        }
                        break;

                    default:
                        return McpUnitySocketHandler.CreateErrorResponse(
                            $"Invalid action '{action}'. Valid actions: 'play', 'pause', 'stop', 'step'",
                            "invalid_parameter"
                        );
                }

                // Give Unity a moment to update state
                bool isPlaying = EditorApplication.isPlaying;
                bool isPaused = EditorApplication.isPaused;

                var result = new JObject
                {
                    ["success"] = true,
                    ["type"] = "text",
                    ["message"] = $"Action '{action}' executed. State: {(isPlaying ? (isPaused ? "Playing (paused)" : "Playing") : "Edit mode")}",
                    ["action"] = action,
                    ["wasPlaying"] = wasPlaying,
                    ["wasPaused"] = wasPaused,
                    ["isPlaying"] = isPlaying,
                    ["isPaused"] = isPaused
                };

                McpLogger.LogInfo($"Play mode action '{action}' executed. isPlaying={isPlaying}, isPaused={isPaused}");

                return result;
            }
            catch (Exception ex)
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    $"Error controlling play mode: {ex.Message}",
                    "play_mode_error"
                );
            }
        }
    }
}
